### PR TITLE
Fix #4829 - endian mach file format

### LINKED ISSUE #4829: Fix endian issue when parsing some files
At least with mach files the magic value gives the information and is saved in bin->big_endian. Use the api `r_read_bleXX` to extract the right value

### PR DESCRIPTION
**Pull request by alvarofe**

### CODE REVIEW

File: libr/bin/format/mach0/mach-o/loader.h [hunk 1]
```diff
@@ -1184,6 +1184,18 @@ struct encryption_info_command {
 				   0 means not-encrypted yet */
 };
 
+
+struct encryption_info_command_64 {
+   ut32	cmd;		/* LC_ENCRYPTION_INFO_64 */
```
Comment (radare) [inline libr/bin/format/mach0/mach-o/loader.h]: wrong indent. use tabs and dont columnize

Comment (radare) [inline libr/bin/format/mach0/mach-o/loader.h]: its src, not xls

File: libr/bin/format/mach0/mach0-swap.c [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+	dyst->indirectsymoff = r_read_ble32 (&dyst->indirectsymoff, endian);
+	dyst->nindirectsyms  = r_read_ble32 (&dyst->nindirectsyms, endian);
+	dyst->extreloff      = r_read_ble32 (&dyst->extreloff, endian);
+	dyst->nextrel 	     = r_read_ble32 (&dyst->nextrel, endian);
+	dyst->locreloff      = r_read_ble32 (&dyst->locreloff, endian);
+	dyst->nlocrel 	     = r_read_ble32 (&dyst->nlocrel, endian);
+}
+
+static void r_mach_swap_dylib_toc(struct dylib_table_of_contents *tocs, ut32 ntocs, bool endian) {
+
```
Comment (radare) [inline libr/bin/format/mach0/mach0-swap.c]: empty line

File: libr/bin/format/mach0/mach0-swap.c [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	for (i = 0; i < ntocs; i++) {
+		tocs[i].symbol_index = r_read_ble32 (&tocs[i].symbol_index, endian);
+		tocs[i].module_index = r_read_ble32 (&tocs[i].module_index, endian);
+	}
+}
+
+static void r_mach_swap_dylib_module_32(struct dylib_module *mods, ut32 nmods, bool endian) {
+	ut32 i;
+	for (i = 0; i < nmods; i++) {
+		mods[i].module_name = r_read_ble32 (&mods[i].module_name, endian);
```
Comment (radare) [inline libr/bin/format/mach0/mach0-swap.c]: why not just use r_buf_fread_at ?

File: libr/bin/format/mach0/mach0.c [hunk 4]
```diff
@@ -84,25 +87,31 @@ static int init_hdr(struct MACH0_(obj_t)* bin) {
 	ut32 magic = 0;
 	int len;
 
-	if (r_buf_read_at (bin->b, 0, (ut8*)&magic, 4) == -1) {
+	if (r_buf_read_at (bin->b, 0, (ut8*)&magic, 4) < 1) {
```
Comment (radare) [inline libr/bin/format/mach0/mach0.c]: use != 4

File: libr/bin/format/mach0/mach0.c [hunk 5]
```diff
@@ -84,25 +87,31 @@ static int init_hdr(struct MACH0_(obj_t)* bin) {
...
-		bin->big_endian?"7I":"7i", 1
-#endif
-	);
-
+	} else if (magic == FAT_MAGIC) {
+	    	bin->big_endian = false;
+	} else {
+	    	return false; // object files are magic == 0, but body is different :?
+	}
+	len = r_buf_read_at (bin->b, 0, (ut8*)&bin->hdr, sizeof(bin->hdr));
```
Comment (radare) [inline libr/bin/format/mach0/mach0.c]: use r_buf_fread_at and provide proper format string

File: libr/bin/format/mach0/mach0.c [hunk 6]
```diff
@@ -125,45 +134,36 @@ static int init_hdr(struct MACH0_(obj_t)* bin) {
...
 		return false;
 	}
-#if R_BIN_MACH064
-	len = r_buf_fread_at (bin->b, off, (ut8*)&bin->segs[seg], bin->big_endian?"2I16c4L4I":"2i16c4l4i", 1);
-#else
-	len = r_buf_fread_at (bin->b, off, (ut8*)&bin->segs[seg], bin->big_endian?"2I16c8I":"2i16c8i", 1);
-#endif
-	if (len < 1)
-		return false;
+	len = r_buf_read_at (bin->b, off, (ut8*)&bin->segs[seg], sizeof(struct MACH0_(segment_command)));
```
Comment (radare) [inline libr/bin/format/mach0/mach0.c]: use r_buf_fread_at

File: libr/bin/format/mach0/mach0.c [hunk 7]
```diff
@@ -580,11 +560,12 @@ static int parse_function_starts (struct MACH0_(obj_t)* bin, ut64 off) {
 			" LC_FUNCTION_STARTS command\n");
 	}
 	bin->func_start = NULL;
-	len = r_buf_fread_at (bin->b, off, (ut8*)&fc, bin->big_endian ? "4I" : "4i", 1);
+	len = r_buf_read_at (bin->b, off, (ut8*)&fc, sizeof(struct linkedit_data_command));
```
Comment (radare) [inline libr/bin/format/mach0/mach0.c]: why removing the use of fread_at ?

File: libr/bin/format/mach0/mach0.c [hunk 8]
```diff
@@ -854,11 +856,12 @@ static int init_items(struct MACH0_(obj_t)* bin) {
 				free (bin->dyld_info);
 				return false;
 			}
-			if (r_buf_fread_at (bin->b, off, (ut8*)bin->dyld_info, bin->big_endian?"12I":"12i", 1) == -1) {
+			if (r_buf_read_at (bin->b, off, (ut8*)bin->dyld_info, sizeof(struct dyld_info_command)) < 1) {
```
Comment (radare) [inline libr/bin/format/mach0/mach0.c]: fread is endian safe

File: libr/bin/format/mach0/mach0.c [hunk 9]
```diff
@@ -895,6 +898,12 @@ static int init_items(struct MACH0_(obj_t)* bin) {
 }
 
 static int init(struct MACH0_(obj_t)* bin) {
+	union {
+		ut16 word;
+		ut8 byte[2];
+	} endian = {1};
```
Comment (radare) [inline libr/bin/format/mach0/mach0.c]: spares around {}

**Comment by alvarofe** [comment]
closing

**Pull request closed without merging**